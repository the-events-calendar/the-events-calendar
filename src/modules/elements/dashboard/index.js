/**
 * Internal dependencies
 */
import Element from './element';

export default Element;
