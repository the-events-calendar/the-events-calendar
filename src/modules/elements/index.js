export { default as Input } from './input';
export { default as TermsList } from './terms-list';
export { default as OrganizerForm } from './organizer-form';
export { default as VenueForm, toFields, toVenue } from './venue-form';
export { default as MetaGroup } from './meta-group';
export { default as SearchPosts } from './search-posts';
export { default as GoogleMap } from './google-map';
export { default as Dashboard } from './dashboard';
export { default as Month } from './month';
export { default as SearchOrCreate } from './search-or-create';
export { default as Loading } from './loading';
export { default as YearMonthForm } from './year-month-form';
export { default as Upsell } from './upsell';
export { default as TimeZone } from './timezone';
export { default as EditLink } from './edit-link';
