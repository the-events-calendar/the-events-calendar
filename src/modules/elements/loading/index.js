import PropTypes from 'prop-types';

/**
 * External dependencies
 */
import React from 'react';
import classNames from 'classnames';

/**
 * WordPress dependencies
 */
import { Spinner } from '@wordpress/components';

import './style.pcss';

/**
 * The loading element.
 *
 * @param {Object} props The component props.
 * @param {string} props.className The class name to add to the container.
 * @returns {JSX.Element} The spinner component.
 */
const Loading = ( { className } ) => {
	return (
		<span className={ classNames( [ 'tribe-editor__spinner-container', className ] ) }>
			<Spinner />
		</span>
	);
};

Loading.propTypes = {
	className: PropTypes.string,
};

export default Loading;
