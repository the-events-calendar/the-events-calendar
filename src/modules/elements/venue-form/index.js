/**
 * Internal dependencies
 */
import Element, { toFields, toVenue } from './element';

export default Element;
export { toFields, toVenue };
