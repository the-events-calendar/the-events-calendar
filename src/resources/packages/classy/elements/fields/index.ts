export * from './EventTitle';
export * from './EventDetails';
export * from './EventDateTime';
export * from './EventOrganizer';
export * from './EventLocation';
