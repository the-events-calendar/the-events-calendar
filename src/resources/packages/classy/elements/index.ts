export * from './Classy';
export * from './fields';
