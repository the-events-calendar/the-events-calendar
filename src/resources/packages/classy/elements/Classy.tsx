import React from 'react';
import { Slot, SlotFillProvider } from '@wordpress/components';
import { doAction } from '@wordpress/hooks';
import { _x } from '@wordpress/i18n';
import { EventDetails, EventTitle } from './fields';
import { RegistryProvider } from '@wordpress/data';
import { WPDataRegistry } from '@wordpress/data/build-types/registry';

function ClassyApplication() {
	return (
		<SlotFillProvider>
			{
				/**
				 * Filters the rendered JSX of the Classy component.
				 *
				 * This component is wrapped within a `SlotFillProvider` to allow dynamic content insertion
				 * via the `Slot/Fill` API. Use the `addFilter` hook to add Fills into the Classy application slots.
				 *
				 * @since TBD
				 */
				doAction( 'classy.render' )
			}

			<div className="classy-container">
				<EventTitle
					title={ _x(
						'Event Title',
						'The title of the event title field.',
						'the-events-calendar'
					) }
				/>

				<EventDetails
					title={ _x(
						'Event Details',
						'The title of the event details field.',
						'the-events-calendar'
					) }
				/>

				{ /* @ts-ignore */ }
				<Slot name="classy.fields" />
			</div>
		</SlotFillProvider>
	);
}

export function Classy( { registry }: { registry: WPDataRegistry } ) {
	{
		/* @ts-ignore */
	}
	return (
		<RegistryProvider value={ registry }>
			<ClassyApplication />
		</RegistryProvider>
	);
}
