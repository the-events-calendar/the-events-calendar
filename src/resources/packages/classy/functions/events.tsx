import { areDatesOnSameDay, areDatesOnSameTime } from '@tec/common/classy/functions';
import { DateTimeUpdateType } from '@tec/common/classy/types/FieldProps';
import { Hours } from '@tec/common/classy/types/Hours';
import { Minutes } from '@tec/common/classy/types/Minutes';
import { NewDatesReturn } from '../types/EventDateTimeDetails';
import { getDurationInDaysForCutoff } from '../functions/date';

/**
 * Calculates new dates for multiday toggle.
 *
 * @since TBD
 *
 * @param {boolean} newValue Whether multiday is being enabled.
 * @param {Date} startDate The current start date.
 * @param {Date} endDate The current end date.
 * @param {Date} defaultStartDate The default start date (8:00 AM).
 * @param {Date} defaultEndDate The default end date (5:00 PM).
 * @param {{start: Date, end: Date} | null} previousDates The previous dates before toggle, if available.
 * @return {{newStartDate: Date, newEndDate: Date}} An object containing the new start and end dates.
 */
export function getMultiDayDates(
	newValue: boolean,
	startDate: Date,
	endDate: Date,
	defaultStartDate: Date,
	defaultEndDate: Date,
	previousDates: { start: Date; end: Date } | null
): { newStartDate: Date; newEndDate: Date } {
	if ( newValue ) {
		// Enable multiday: start date + 24 hours + duration difference
		const duration = endDate.getTime() - startDate.getTime();
		const newEndDate = new Date( startDate.getTime() + 24 * 60 * 60 * 1000 + duration );
		return { newStartDate: startDate, newEndDate };
	} else {
		// Disable multiday: revert to previous state if available, otherwise default
		const revertStart = previousDates ? previousDates.start : defaultStartDate;
		const revertEnd = previousDates ? previousDates.end : defaultEndDate;
		return { newStartDate: revertStart, newEndDate: revertEnd };
	}
}

/**
 * Calculates new start and end dates when toggling all-day events.
 *
 * @since TBD
 *
 * @param {boolean} newValue Indicates whether the event is now an all-day event.
 * @param {Date} startDate The current start date of the event.
 * @param {{hours: Hours, minutes: Minutes}} endOfDayCutoff The time at which the day ends.
 * @param {Date} endDate The current end date of the event.
 * @param {Date} defaultStartDate The default start date (8:00 AM).
 * @param {Date} defaultEndDate The default end date (5:00 PM).
 * @param {{start: Date, end: Date} | null} previousDates The previous dates before toggle, if available.
 * @return {{newStartDate: Date, newEndDate: Date}} An object containing the new start and end dates.
 */
export function getAllDayNewDates(
	newValue: boolean,
	startDate: Date,
	endOfDayCutoff: {
		hours: Hours;
		minutes: Minutes;
	},
	endDate: Date,
	defaultStartDate: Date,
	defaultEndDate: Date,
	previousDates: { start: Date; end: Date } | null
): { newStartDate: Date; newEndDate: Date } {
	if ( newValue ) {
		// Enable all-day: set to full day.
		const newStartDate = new Date( startDate );
		newStartDate.setHours( endOfDayCutoff.hours, endOfDayCutoff.minutes, 0, 0 );
		const durationInDays = getDurationInDaysForCutoff( endOfDayCutoff, startDate, endDate );
		// The event will last at least one day and the cutoff minus one second.
		const duration = ( durationInDays + 1 ) * 24 * 60 * 60 * 1000 - 1;
		const newEndDate = new Date( newStartDate.getTime() + duration );

		return { newStartDate, newEndDate };
	} else {
		// Disable all-day: revert to previous state if available, otherwise default.
		const revertStart = previousDates ? previousDates.start : defaultStartDate;
		const revertEnd = previousDates ? previousDates.end : defaultEndDate;
		return { newStartDate: revertStart, newEndDate: revertEnd };
	}
}

/**
 * Calculates new start and end dates based on user updates.
 *
 * @since TBD
 *
 * @param {Date} endDate The current end date.
 * @param {Date} startDate The current start date.
 * @param {'start' | 'end'} updated Indicates whether the start or end date was updated.
 * @param {string} newDate The new date string provided by the user.
 * @param {boolean} isMultiDayEnabled Whether multiday is enabled.
 *
 * @return {NewDatesReturn} An object defining the new start and end dates, and whether the user needs to be notified
 *     of the implicit change of either.
 */
export function getNewStartEndDates(
	endDate: Date,
	startDate: Date,
	updated: DateTimeUpdateType,
	newDate: string,
	isMultiDayEnabled: boolean
): NewDatesReturn {
	// Milliseconds.
	const duration = endDate.getTime() - startDate.getTime();

	// By default, do not move the start date but keep it to the previous value.
	let newStartDate = startDate;

	// By default, do not move the end date but keep it to the previous value.
	let newEndDate = endDate;
	let notify = { startDate: false, startTime: false, endDate: false, endTime: false };

	try {
		switch ( updated ) {
			case 'startDate':
				newStartDate = new Date( newDate );

				// If not multiday update end date with original duration.
				if ( ! isMultiDayEnabled ) {
					newEndDate = new Date( newDate );
					newEndDate.setHours( endDate.getHours(), endDate.getMinutes() );
				}

				break;
			case 'startTime':
				// The user has updated the start date.
				newStartDate = new Date( newDate );

				if ( newStartDate.getTime() >= endDate.getTime() ) {
					// For time updates, push end time to next interval
					newEndDate = new Date( newStartDate );
				}

				break;
			case 'endDate':
				// The user has updated the end date.
				newEndDate = new Date( newDate );
				if ( newEndDate.getTime() <= startDate.getTime() ) {
					// For date updates, maintain duration
					newStartDate = new Date( newEndDate.getTime() - duration );
				}

				break;
			case 'endTime':
				// The user has updated the end date.
				newEndDate = new Date( newDate );

				if ( newEndDate.getTime() < startDate.getTime() ) {
					// For time updates, pull start time to previous interval.
					newStartDate = new Date( newEndDate.getTime() - duration );
				}
				break;
		}

		// Highlight the start date if it actually changed as a consequence of the update.
		notify.startDate = updated !== 'startDate' && ! areDatesOnSameDay( startDate, newStartDate );

		// Highlight the start time if it actually changed as a consequence of the update.
		notify.startTime = updated !== 'startTime' && ! areDatesOnSameTime( startDate, newStartDate );

		// Highlight the end date if it actually changed as a consequence of the update.
		notify.endDate = updated !== 'endDate' && ! areDatesOnSameDay( endDate, newEndDate );

		// Highlight the end time if it actually changed as a consequence of the update.
		notify.endTime = updated !== 'endTime' && ! areDatesOnSameTime( endDate, newEndDate );
	} catch ( e ) {
		// Something went wrong while processing the dates, return the values unchanged and notify no field.
		newStartDate = startDate;
		newEndDate = endDate;
		// Nothing to notify since nothing changed.
		notify = { startDate: false, startTime: false, endDate: false, endTime: false };
	}

	return { newStartDate, newEndDate, notify };
}
