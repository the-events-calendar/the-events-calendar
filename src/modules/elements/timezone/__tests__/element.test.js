/**
 * External dependencies
 */
import React from 'react';
import renderer from 'react-test-renderer';

/**
 * Internal dependencies
 */
import { TimeZone } from '@moderntribe/events/elements';

describe( 'TimeZone element', () => {
	it( 'Should render the component', () => {
		const component = renderer.create( <TimeZone /> );
		expect( component.toJSON() ).toMatchSnapshot();
	} );

	it( 'Should render the value on the input', () => {
		const component = renderer.create( <TimeZone value="Modern Tribe" /> );
		expect( component.toJSON() ).toMatchSnapshot();
	} );

	it( 'Should render the placeholder on the input', () => {
		const component = renderer.create( <TimeZone placeholder="Love this tribe ! " /> );
		expect( component.toJSON() ).toMatchSnapshot();
	} );

	it( 'Should add the class name into the component', () => {
		const component = renderer.create( <TimeZone className="my-custom-class-name" /> );
		expect( component.toJSON() ).toMatchSnapshot();
	} );

	it( 'Should trigger the onChange event', () => {
		const onChange = jest.fn();
		const component = renderer.create( <TimeZone value="Modern Tribe" onChange={ onChange } /> );
		const tree = component.toJSON();

		// Find the input element and trigger its onChange handler with a synthetic event
		const input = tree.children.find( ( child ) => child.type === 'input' );
		if ( input && input.props.onChange ) {
			input.props.onChange( { target: { value: 'Modern Tribe' } } );
		}

		expect( onChange ).toHaveBeenCalled();
		expect( onChange ).toHaveBeenCalledWith( 'Modern Tribe' );
	} );
} );
