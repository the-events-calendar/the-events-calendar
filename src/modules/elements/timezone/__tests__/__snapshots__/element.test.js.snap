// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeZone element Should add the class name into the component 1`] = `
<div
  className="my-custom-class-name"
  style={
    {
      "display": "inline-block",
    }
  }
>
  <input
    onChange={[Function]}
    placeholder=""
    style={
      {
        "boxSizing": "content-box",
        "width": "1px",
      }
    }
    value=""
  />
  <div
    style={
      {
        "height": 0,
        "left": 0,
        "overflow": "scroll",
        "position": "absolute",
        "top": 0,
        "visibility": "hidden",
        "whiteSpace": "pre",
      }
    }
  />
</div>
`;

exports[`TimeZone element Should render the component 1`] = `
<div
  className="tribe-editor__timezone-input"
  style={
    {
      "display": "inline-block",
    }
  }
>
  <input
    onChange={[Function]}
    placeholder=""
    style={
      {
        "boxSizing": "content-box",
        "width": "1px",
      }
    }
    value=""
  />
  <div
    style={
      {
        "height": 0,
        "left": 0,
        "overflow": "scroll",
        "position": "absolute",
        "top": 0,
        "visibility": "hidden",
        "whiteSpace": "pre",
      }
    }
  />
</div>
`;

exports[`TimeZone element Should render the placeholder on the input 1`] = `
<div
  className="tribe-editor__timezone-input"
  style={
    {
      "display": "inline-block",
    }
  }
>
  <input
    onChange={[Function]}
    placeholder="Love this tribe ! "
    style={
      {
        "boxSizing": "content-box",
        "width": "1px",
      }
    }
    value=""
  />
  <div
    style={
      {
        "height": 0,
        "left": 0,
        "overflow": "scroll",
        "position": "absolute",
        "top": 0,
        "visibility": "hidden",
        "whiteSpace": "pre",
      }
    }
  />
  <div
    style={
      {
        "height": 0,
        "left": 0,
        "overflow": "scroll",
        "position": "absolute",
        "top": 0,
        "visibility": "hidden",
        "whiteSpace": "pre",
      }
    }
  >
    Love this tribe ! 
  </div>
</div>
`;

exports[`TimeZone element Should render the value on the input 1`] = `
<div
  className="tribe-editor__timezone-input"
  style={
    {
      "display": "inline-block",
    }
  }
>
  <input
    onChange={[Function]}
    placeholder=""
    style={
      {
        "boxSizing": "content-box",
        "width": "1px",
      }
    }
    value="Modern Tribe"
  />
  <div
    style={
      {
        "height": 0,
        "left": 0,
        "overflow": "scroll",
        "position": "absolute",
        "top": 0,
        "visibility": "hidden",
        "whiteSpace": "pre",
      }
    }
  >
    Modern Tribe
  </div>
</div>
`;
